File: python/src/coding_interview/infrastructure/repository/portfolio_repository_impl.py
from __future__ import annotations

from coding_interview.domain.constants import INITIAL_PORTFOLIO
from coding_interview.domain.stock import Portfolio


class PortfolioRepositoryImpl:
    def __init__(self) -> None:
        self._portfolio: Portfolio = INITIAL_PORTFOLIO

    def get(self) -> Portfolio:
        return self._portfolio

    def update(self, portfolio: Portfolio) -> None:
        self._portfolio = portfolio

File: ruby/lib/coding_interview/infrastructure/repository/portfolio_repository_impl.rb
require_relative "../../application/repository/portfolio_repository"
require_relative "../../domain/app_constants"

module CodingInterview
  module Infrastructure
    module Repository
      class PortfolioRepositoryImpl < Application::Repository::PortfolioRepository
        def initialize
          @portfolio = Domain::AppConstants::INITIAL_PORTFOLIO
          @mutex = Mutex.new
        end

        def get
          @mutex.synchronize { @portfolio }
        end

        def update(portfolio)
          @mutex.synchronize { @portfolio = portfolio }
          nil
        end
      end
    end
  end
end
